scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1749 @@
+ *
+ * Intuitively, if Task B's runtime per schedule is long, a relatively short
+ * scheduling delay won't affect a lot; if Task B frequently wakes up Task C,
+ * the scheduling delay of Task B also delays the execution of Task C;
+ * similarly, if Task B often waits for Task A, the scheduling delay of Task B
+ * delays the completion of executing the task graph.
+ *
+ *
+ * 3. Virtual deadline: when to execute a task
+ * -------------------------------------------
+ *
+ * The latency criticality of a task is used to determine task's virtual
+ * deadline. A more latency-critical task will have a tighter (shorter)
+ * deadline, so the scheduler picks such a task more urgently among runnable
+ * tasks.
+ *
+ *
+ * 4. Time slice: how long execute a task
+ * --------------------------------------
+ *
+ * We borrow the time slice calculation idea from the CFS and scx_rustland
+ * schedulers. The LAVD scheduler tries to schedule all the runnable tasks at
+ * least once within a predefined time window, which is called a targeted
+ * latency. For example, if a targeted latency is 15 msec and 10 tasks are
+ * runnable, the scheduler equally divides 15 msec of CPU time into 10 tasks.
+ * Of course, the scheduler will consider the task's priority -- a task with
+ * higher priority (lower nice value) will receive a longer time slice.
+ *
+ * The scheduler also considers the behavioral properties of a task in
+ * determining the time slice. If a task is compute-intensive, so it consumes
+ * the assigned time slice entirely, the scheduler boosts such task's time
+ * slice and assigns a longer time slice. Next, if a task is freshly forked,
+ * the scheduler assigns only half of a regular time slice so it can make a
+ * more educated decision after collecting the behavior of a new task. This
+ * helps to mitigate fork-bomb attacks.
+ *
+ *
+ * 5. Fairness: how to enforce the fair use of CPU time
+ * ----------------------------------------------------
+ *
+ * Assigning a task's time slice per its priority does not guarantee the fair
+ * use of CPU time. That is because a task can be more (or less) frequently
+ * executed than other tasks or yield CPU before entirely consuming its
+ * assigned time slice.
+ *
+ * The scheduler treats the over-scheduled (or ineligible) tasks to enforce the
+ * fair use of CPU time. It defers choosing over-scheduled tasks to reduce the
+ * frequency of task execution. The deferring time- ineligible duration- is
+ * proportional to how much time is over-spent and added to the task's
+ * deadline. Additionally, suppose a task is compute-intensive and not
+ * latency-critical tasks. In that case, so its runtime per schedule is long
+ * enough without voluntarily yielding CPU time, the scheduler reduces the time
+ * slice, too. Note that reducing the time slice of a latency-critical task for
+ * fairness is not very effective because the scheduling overhead might be
+ * detrimental.
```

Comment:
This last part about fairness is not very clear to me. Let me try to rephrase to make sure I understand:

Additionally, if a task is compute-intensive and not latency-critical, the scheduler automatically reduces its time slice, since its runtime per schedule is sufficiently long enough without voluntarily yielding the CPU.

Is this the same concept?

Reply:
Your version is better. I will change the last paragraph as follows:

 * The scheduler treats the over-scheduled (or ineligible) tasks to enforce the
 * fair use of CPU time. It defers choosing over-scheduled tasks to reduce the
 * frequency of task execution. The deferring time- ineligible duration- is
 * proportional to how much time is over-spent and added to the task's
 * deadline. Additionally, if a task is compute-intensive and not
 * latency-critical, the scheduler automatically reduces its time slice, since
 * its runtime per schedule is sufficiently long enough without voluntarily
 * yielding the CPU. Note that reducing the time slice of a latency-critical
 * task for fairness is not very effective because the scheduling overhead
 * might be detrimental.